src/physics/src/spalart_allmaras_helper.C

```diff
@@ -0,0 +1,152 @@
+// Public License as published by the Free Software Foundation.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc. 51 Franklin Street, Fifth Floor,
+// Boston, MA  02110-1301  USA
+//
+//-----------------------------------------------------------------------el-
+
+
+// This class
+#include "grins/spalart_allmaras_helper.h"
+
+// GRINS
+#include "grins/assembly_context.h"
+#include "grins/generic_ic_handler.h"
+#include "grins/inc_navier_stokes_bc_handling.h"
+#include "grins/turbulence_models_macro.h"
+
+// libMesh
+#include "libmesh/quadrature.h"
+#include "libmesh/elem.h"
+#include "libmesh/unstructured_mesh.h"
+#include "libmesh/fem_system.h"
+
+namespace GRINS
+{
+  
+  SpalartAllmarasHelper::SpalartAllmarasHelper(const GetPot& input )
+    : _cb1(0.1355), // Define class variables
+      _sigma(2./3.),
+      _cb2(0.622),
+      _kappa(0.41),
+      _cv1(7.1),
+      _cv2(0.7),
+      _cv3(0.9),
+      _r_lin(10.0),
+      _c_w2(0.3),
+      _c_w3(2.0),
+      _c_t3(1.2),
+      _c_t4(0.5),
+      _c_n1(16.0),
+      _flow_vars(input)
+  {        
+    _cw1 = _cb1/pow(_kappa,2.0) + (1 + _cb2)/_sigma;
+
+    return;
+  }
+  
+  SpalartAllmarasHelper::~SpalartAllmarasHelper()
+  {
+    return;
+  }
+ 
+  void SpalartAllmarasHelper::init_variables( libMesh::FEMSystem* system )
+  {    
+    this->_dim = system->get_mesh().mesh_dimension();
+
+    this->_flow_vars.init(system);
+        
+    return;
+  }
+    
+  libMesh::Real SpalartAllmarasHelper::_vorticity(AssemblyContext& context, unsigned int qp) const
+  {
+    libMesh::Gradient grad_u, grad_v;
+    grad_u = context.interior_gradient(this->_flow_vars.u_var(), qp);
+    grad_v = context.interior_gradient(this->_flow_vars.v_var(), qp);
+
+    libMesh::Real _vorticity_value;
+    _vorticity_value = fabs(grad_v(0) - grad_u(1));
+
+    if(this->_dim == 3)
+      {
+	libMesh::Gradient grad_w;
+	grad_w = context.interior_gradient(this->_flow_vars.w_var(), qp);
+	
+	libMesh::Real _vorticity_component_0 = grad_w(1) - grad_v(2);
+	libMesh::Real _vorticity_component_1 = grad_u(2) - grad_v(0);
+
+	_vorticity_value += pow(pow(_vorticity_component_0, 2.0) + pow(_vorticity_component_1, 2.0) + pow(_vorticity_value, 2.0), 0.5);
+      }
+
+    return _vorticity_value;
+  }
+  
+  libMesh::Real SpalartAllmarasHelper::_source_fn(libMesh::Number nu, libMesh::Real mu, libMesh::Real wall_distance, libMesh::Real _vorticity_value) const
+  {
+    // Step 1
+    libMesh::Real _chi = nu/mu;
+    
+    // Step 2
+    libMesh::Real _fv1 = pow(_chi, 3.0)/(pow(_chi, 3.0) + pow(this->_cv1, 3.0));
```

Comment:
If you use "3.0" instead of "3" here then I'm not sure even compilers with pow(foo,int) specializations are going to be smart enough to reduce that down to "_chi*_chi*_chi".